### PR TITLE
Teach meshmode about FEM program transformation

### PR DESCRIPTION
**Pull request by inducer**
This supersedes https://github.com/inducer/arraycontext/pull/29, in that it never made sense that `arraycontext` had to know anything about how to transform FEM-related programs. Together with https://github.com/inducer/arraycontext/pull/50, this PR fixes that.

This PR is probably best read commit-by-commit.

If you'd like to run this, you'll need https://github.com/inducer/loopy/pull/445 (already merged at the time of this writing).

Draft (but ready for review) because:
- [x] Needs https://github.com/inducer/arraycontext/pull/50
- [x] `requirements.txt` needs to be redangled.

### CODE REVIEW

**Comment by inducer** [comment]
(I think the Firedrake thing is easily solved... but not tonight. :)

File: test/test_array.py [hunk 1]
```diff
@@ -24,10 +24,10 @@
 import pytest
 import numpy as np
 
-from arraycontext import _acf           # noqa: F401
-from arraycontext import (              # noqa: F401
-        pytest_generate_tests_for_pyopencl_array_context
-        as pytest_generate_tests)
+from meshmode.array_context import PytestPyOpenCLArrayContextFactory
```
Comment (alexfikl) [inline test/test_array.py:27]:
```suggestion
from meshmode.array_context import PytestPyOpenCLArrayContextFactory, _acf
```
? Or update the one in `meshmode._acf`?

File: test/test_interop.py [hunk 2]
```diff
@@ -24,10 +24,12 @@
 import numpy as np
 import pytest
 
-from arraycontext import PyOpenCLArrayContext, thaw, _acf   # noqa: F401
-from arraycontext import (                                  # noqa: F401
-        pytest_generate_tests_for_pyopencl_array_context
-        as pytest_generate_tests)
+from arraycontext import PyOpenCLArrayContext, thaw   # noqa: F401
```
Comment (alexfikl) [inline test/test_interop.py:27]:
```suggestion
from arraycontext import thaw
```

File: test/test_modal.py [hunk 3]
```diff
@@ -24,11 +24,11 @@
 from functools import partial
 import numpy as np
 
-from arraycontext import thaw, _acf     # noqa: F401
-from arraycontext import (              # noqa: F401
-    pytest_generate_tests_for_pyopencl_array_context
-    as pytest_generate_tests
-    )
+from arraycontext import thaw     # noqa: F401
```
Comment (alexfikl) [inline test/test_modal.py:27]:
```suggestion
from arraycontext import thaw
```

File: test/test_partition.py [hunk 4]
```diff
@@ -28,10 +28,11 @@
 
 from meshmode.dof_array import flatten, unflatten, flat_norm
 
-from arraycontext import thaw, _acf         # noqa: F401
-from arraycontext import (                  # noqa: F401
-        pytest_generate_tests_for_pyopencl_array_context
-        as pytest_generate_tests)
+from arraycontext import thaw         # noqa: F401
```
Comment (alexfikl) [inline test/test_partition.py:31]:
```suggestion
from arraycontext import thaw
```

File: test/test_refinement.py [hunk 5]
```diff
@@ -26,10 +26,12 @@
 import numpy as np
 import pytest
 
-from arraycontext import thaw, _acf         # noqa: F401
-from arraycontext import (                  # noqa: F401
-        pytest_generate_tests_for_pyopencl_array_context
-        as pytest_generate_tests)
+from arraycontext import thaw       # noqa: F401
```
Comment (alexfikl) [inline test/test_refinement.py:29]:
```suggestion
from arraycontext import thaw
```

File: test/test_visualization.py [hunk 6]
```diff
@@ -39,10 +39,11 @@
         )
 import meshmode.mesh.generation as mgen
 
-from arraycontext import thaw, _acf         # noqa: F401
-from arraycontext import (                  # noqa: F401
-        pytest_generate_tests_for_pyopencl_array_context
-        as pytest_generate_tests)
+from arraycontext import thaw      # noqa: F401
```
Comment (alexfikl) [inline test/test_visualization.py:42]:
```suggestion
from arraycontext import thaw
```

**Review by alexfikl — APPROVED** [review]
Can't comment too much on the transform logic (didn't spot anything fishy), so this looks good to me!

File: test/test_array.py (hunk 1 above)
Comment (inducer) [reply test/test_array.py:28]: 130745984ef3eb193df0128dedd250ef07d35160

File: test/test_visualization.py (hunk 6 above)
Comment (inducer) [reply test/test_visualization.py]: Done as part of this force-push: https://github.com/inducer/meshmode/compare/c633878535ce4d1b0cd397fefcafef9abf8e6465..130745984ef3eb193df0128dedd250ef07d35160

File: test/test_refinement.py (hunk 5 above)
Comment (inducer) [reply test/test_refinement.py]: Done as part of this force-push: https://github.com/inducer/meshmode/compare/c633878535ce4d1b0cd397fefcafef9abf8e6465..130745984ef3eb193df0128dedd250ef07d35160

File: test/test_partition.py (hunk 4 above)
Comment (inducer) [reply test/test_partition.py]: Done as part of this force-push: https://github.com/inducer/meshmode/compare/c633878535ce4d1b0cd397fefcafef9abf8e6465..130745984ef3eb193df0128dedd250ef07d35160

File: test/test_modal.py (hunk 3 above)
Comment (inducer) [reply test/test_modal.py]: Done as part of this force-push: https://github.com/inducer/meshmode/compare/c633878535ce4d1b0cd397fefcafef9abf8e6465..130745984ef3eb193df0128dedd250ef07d35160

File: test/test_interop.py (hunk 2 above)
Comment (inducer) [reply test/test_interop.py]: Done as part of this force-push: https://github.com/inducer/meshmode/compare/c633878535ce4d1b0cd397fefcafef9abf8e6465..130745984ef3eb193df0128dedd250ef07d35160

File: meshmode/array_context.py [hunk 7]
```diff
@@ -57,7 +38,266 @@ def thaw(actx, ary):
...
+                    if assignee.aggregate.name not in first_axis_el_args:
+                        continue
+
+                    subscripts = assignee.index_tuple[:2]
+
+                    for i, subscript in enumerate(subscripts):
+                        if (not isinstance(subscript, Variable)
+                                or subscript.name not in default_ep.all_inames()):
+                            raise ValueError("subscripts in "
+                                    "ElementwiseMapKernelTag-tagged kernels must be "
```
[the remark this answers is not in the document]
Comment (inducer) [reply meshmode/array_context.py]: Done as part of this force-push: https://github.com/inducer/meshmode/compare/c633878535ce4d1b0cd397fefcafef9abf8e6465..130745984ef3eb193df0128dedd250ef07d35160

File: meshmode/array_context.py [hunk 8]
```diff
@@ -57,7 +38,266 @@ def thaw(actx, ary):
...
+    if first_axis_el_args:
+        el_inames = []
+        dof_inames = []
+
+        for stmt in default_ep.instructions:
+            if isinstance(stmt, lp.MultiAssignmentBase):
+                for assignee in stmt.assignees:
+                    if not isinstance(assignee, Subscript):
+                        raise ValueError("assignees in "
+                                "ElementwiseMapKernelTag-tagged kernels must be "
```
[the remark this answers is not in the document]
Comment (inducer) [reply meshmode/array_context.py]: Done as part of this force-push: https://github.com/inducer/meshmode/compare/c633878535ce4d1b0cd397fefcafef9abf8e6465..130745984ef3eb193df0128dedd250ef07d35160

**Comment by inducer** [comment]
Thanks for taking a look!

**Pull request merged**